### PR TITLE
Enhancements and bugs fixes.

### PR DESCRIPTION
**Pull request by jui8wang**
1. Support getCommandType() in Imap command class.
2. Support ImapResponseParser to parse response from SearchCommand and UidSearchCommand to appropriate object.
3. Support MessageNumberSet to allow converting int or long array to its own, also supports last message
4. Fix bug: ImapResponseMapper in parsing to AppendUid.
5. Fix bug: Uid Search command to pass message sequences instead of UID.
6. Fix bug: fetch macro should not enclose parenthesis
7. Fix bug: store flags command with FlagsAction enum (replace, add, remove)
8. Fix bug: login command parameters need encoding
9. Fix bug: Auth plain with auth id
10. Support Append with Literal+.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Major release (change is NOT backward compatible with prior release)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    /**
+     * Converts an array of long into array of MessageNumberSet. This is a helper method for callers with long array.
+     *
+     * @param msgs array of long data type message number (could be message sequence or UID)
+     * @return the string generated that conforms to RFC3501 Message Number syntax
+     */
+    public static MessageNumberSet[] createMessageNumberSets(@Nonnull final long[] msgs) {
+        List<MessageNumberSet> v = new ArrayList<MessageNumberSet>();
+        int i, j;
+        for (i = 0; i < msgs.length; i++) {
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java]: You know that i++ is optional

File: core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+            long start = msgs[i];
+
+            // Look for contiguous elements
+            for (j = i + 1; j < msgs.length; j++) {
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java]: Would these look better as a while loop?

File: core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+        for (;;) {
+            start = msgsets[i].start;
+            end = msgsets[i].end;
+
+            if (msgsets[i].seqType == SequenceType.LAST_MESSAGE_ONLY) {
+                s.append('*');
+            } else if (msgsets[i].seqType == SequenceType.LAST_MESSAGE_END) {
+                s.append(start).append(':').append('*');
+            } else if (end > start) {
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java]: In imap, sequence range can be defined 1:10 or 10:1. Both are equivalent.

File: core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        if (msgsets == null || msgsets.length == 0) {
+            return null;
+        }
+
+        int i = 0; // msgset index
+        final StringBuilder s = new StringBuilder();
+        int size = msgsets.length;
+        long start, end;
+
+        for (;;) {
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java]: Infinite loops are not good. I would have a terminating condition.

File: core/src/main/java/com/yahoo/imapnio/async/internal/CompressCommand.java [hunk 5]
```diff
@@ -13,12 +14,16 @@
...
 
     /**
-     * Initializes the @{code COMPRESS_DEFLATECommand}. Constructor is only visible to this package so it can only be called by @{code
-     * ImapAsyncSessionImpl}.
+     * Initializes the @{code CompressCommand}. Constructor is only visible to this package so it can only be called by @{code ImapAsyncSessionImpl}.
      * 
      * @see "RFC 4978"
      */
     CompressCommand() {
         super(COMPRESS_DEFLATE);
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/internal/CompressCommand.java]: What is the difference between this and IMAPCommandType?

File: core/src/main/java/com/yahoo/imapnio/async/request/AbstractFetchCommand.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    private static final String FETCH_SPACE = "FETCH ";
+
+    /** FETCH and space. */
+    private static final String UID_FETCH_SPACE = "UID FETCH ";
+
+    /** Message numbers, either message sequence or UID. */
+    private String msgNumbers;
+
+    /** Fetch items. */
+    private String dataItems;
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/request/AbstractFetchCommand.java]: Should we have separate classes one with message data items and another with macro?

File: core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java (hunk 4 above)
Comment (jui8wang) [reply core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java]: Done

File: core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java (hunk 1 above)
Comment (jui8wang) [reply core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java]: Done

File: core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java (hunk 2 above)
Comment (jui8wang) [reply core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java]: Done.

File: core/src/main/java/com/yahoo/imapnio/async/internal/CompressCommand.java (hunk 5 above)
Comment (jui8wang) [reply core/src/main/java/com/yahoo/imapnio/async/internal/CompressCommand.java]: that is the command string, for example, "COMPRESS DEFLATE", that is used for the command sent to server.  It is not an enum.

File: core/src/main/java/com/yahoo/imapnio/async/request/AbstractFetchCommand.java (hunk 6 above)
Comment (jui8wang) [reply core/src/main/java/com/yahoo/imapnio/async/request/AbstractFetchCommand.java]: That would end up 2*2  classes (FetchMacroCommand,  FetchDataItemsCommand, UidFetchMacroCommand, UidFetchDataItemsCommand),  it is going to be a little too much to maintain.

File: core/src/main/java/com/yahoo/imapnio/async/request/ImapClientConstants.java [hunk 7]
```diff
@@ -12,6 +11,9 @@
     /** String for CR and LF. */
     public static final String CRLF = "\r\n";
 
+    /** Byte array for CR and LF. */
+    public static final byte[] CRLF_B = { '\r', '\n' };
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/request/ImapClientConstants.java]: fyi, arrays are bad for exposing as public variables. You should probably store as ByteBuf objects.

File: core/src/main/java/com/yahoo/imapnio/async/request/AbstractSearchCommand.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    private String msgNumbers;
+
+    /** The search expression. */
+    private Argument searchExpr;
+
+    /** Character set. */
+    private String charset;
+
+    /** flag whether server allows LITERAL+. */
+    private boolean isLiteralPlusEnabled;
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/request/AbstractSearchCommand.java]: Hmm. Fyi, this is something related to data streaming. Should this be included in request?

File: core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java (hunk 2 above)
Comment (jui8wang) [reply core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java]: yours is better.  I changed.

File: core/src/main/java/com/yahoo/imapnio/async/request/ImapClientConstants.java (hunk 7 above)
Comment (jui8wang) [reply core/src/main/java/com/yahoo/imapnio/async/request/ImapClientConstants.java]: Done.  Good point!

Comment (jui8wang) [reply core/src/main/java/com/yahoo/imapnio/async/request/ImapClientConstants.java]: After discussing regarding the constraints of ByteBuf or ByteBuffer,  I have moved this array private to individual classes.

**Comment by nvsnvikram** [comment]
I have one command on the list folder. Otherwise, this seems to be good.

File: core/src/main/java/com/yahoo/imapnio/async/request/AbstractQueryFoldersCommand.java [hunk 9]
```diff
@@ -39,24 +48,25 @@ public void cleanup() {
     }
 
     @Override
-    public String getCommandLine() {
+    public ByteBuf getCommandLineBytes() throws ImapAsyncClientException {
         // Ex:LIST /usr/staff/jones ""
 
         // encode the arguments as per RFC2060
         final String ref64 = BASE64MailboxEncoder.encode(ref);
         final String pat64 = BASE64MailboxEncoder.encode(pattern);
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/request/AbstractQueryFoldersCommand.java]: I remember according to imap rfc, this does not have to be imap utf-7 encoded. Please check.

**Pull request merged**